商品購入機能

sh-o-kada opened the pull request:
What：商品購入ページ、クレジットカード決済機能の実装

Why：商品購入機能実装のため

gyazo
・[必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画](https://gyazo.com/e118a6c8313fe5e2a250285264b64e05)

・[入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画](https://gyazo.com/b1c9fee9d943181c4eecb32c303659a0)

・[ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画](https://gyazo.com/9e989596cd960a33f03c087cc5b0c06f)

・[ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画](https://gyazo.com/08c7a27e1c4376628a5f7c0cbf25d04e)

・[ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画](https://gyazo.com/4ee37cd692fd872fab9e0eb592f446d2)

・[売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）](https://gyazo.com/546445c4edb36ae18ca7eab3201cafb8)

・[売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品詳細機能実装時に未実装であった場合）](https://gyazo.com/5be45ae61d351112bbec1137cb2fdc23)

・[ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）](https://gyazo.com/c0ed0152273616f24c84d597ecbd9e41)

・[ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）](https://gyazo.com/e76da11e647a641747b92e7ac47cabe5)

・[テスト結果の画像](https://gyazo.com/f2e0115f26acc333896d27dec6dd2519)

sh-o-kada committed:
商品購入ページの実装

商品購入ページの作成
商品購入情報保存機能の実装
商品購入(Orderモデル)単体テストコードの作成

sh-o-kada committed:
売却済み商品の表示の実装

sh-o-kada committed:
クレジットカード決済機能の実装

zonodance commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,42 @@
+  before_action :authenticate_user!, only: [:index]
+
+  def index
+    @item = Item.find(params[:item_id])
```
`@item = Item.find(params[:item_id])`
createアクションにも同様の記述があるので、before_actionでまとめてみましょう。

zonodance commented on app/models/prefecture.rb:
```diff
@@ -21,4 +21,5 @@ class Prefecture < ActiveHash::Base
 
   include ActiveHash::Associations
   has_many :items
+  has_many :adresses
```
`has_many :adresses`
正しいモデルをアソシエーションで指定しましょう。

zonodance commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,79 @@
+        @order_address.valid?
+        expect(@order_address.errors.full_messages).to include("Phone can't be blank")
+      end
+      it 'phoneが10桁or11桁の半角数字じゃないと保存できない' do
```
電話番号のテストは、テストコードの整合性を高めるために、2つのテストに分けて実装しましょう！
・電話番号が9桁以下では購入できない
・電話番号が12桁以上では購入できない

zonodance commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,79 @@
+        @order_address.phone = '123456789'
+        @order_address.valid?
+        expect(@order_address.errors.full_messages).to include('Phone には10桁or11桁の半角数字で入力してください')
+      end
```
電話番号のテストは、テストコードの整合性を高めるために、数値以外の内容が記述された場合のパターンも検討しましょう。
・電話番号に半角数字以外が含まれている場合は購入できない

zonodance reviewed (commented):
レビュー依頼ありがとうございます。
コメント内容を確認の上再度実装に取り組んでみましょう。

sh-o-kada committed:
商品購入機能の修正1

・app/models/prefecture.rb、アソシエーションの修正
・app/controllers/orders_controller.rb、リファクタリング
・spec/models/order_address_spec.rb、phoneのテストの場合分け

sh-o-kada commented:
ご確認ありがとうございます。修正致しました。
[修正後のテスト結果](https://gyazo.com/656f4856d764aadddc53c588e60a39eb)です。

zonodance reviewed (commented):
レビュー依頼ありがとうございます！

`LGTM`とさせていただきます！商品購入機能の実装お疲れさまでした。
次のブランチを切る前にMergeとPullを忘れないようお願いします！

sh-o-kada commented:
ご対応ありがとうございました！
